core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java

```diff
@@ -987,7 +987,7 @@ public SqlNode field(int ordinal) {
       } else {
         boolean qualified =
             !dialect.hasImplicitTableAlias() || aliases.size() > 1;
-        if (needNew) {
+        if (needNew && !aliases.containsKey(neededAlias)) {
```

Comment:
I think we should be careful here and at least check that the aliases is of size 1 and that we have a new "neededAlias".  One thing to be careful here is that "neededAlias" can be null.